Add tag_prefix config option

vortechs2000 opened the pull request:
tag_prefix lets one specify the prefix to the actual tag. This is helpful in monorepos where many product tags exist, and there is a consistent prefix for the tags to be used.

Potential fix for #68

dolfinus commented:
Hi, thank you for your contribution!

I see here 2 ways to implement that:
1. Pass prefix directly to `git tag -l` command (but this is not a regexp):
```
OPTIONS
       -l, --list
           List tags. With optional <pattern>..., e.g.  git tag --list 'v-*', list only the tags that match the pattern(s).

           Running "git tag" without arguments also lists all tags. The pattern is a shell wildcard (i.e., matched using fnmatch(3)). Multiple patterns may be given; if any of them matches, the tag is shown.
```
2. Add `tag_filter` option which could be a regexp or a callback (like `tag_formatter` and `branch_formatter`), which allows users to discard tags not only by prefix, but by any logic they could implement

Also I'm not sure about implicit change of default value of `tag_formatter` according to `tag_prefix`.

vortechs2000 commented:
Squashed the history down to make this a bit cleaner - just the `tag_filter` changes in the commits now

dolfinus commented on setuptools_git_versioning.py:
```diff
@@ -500,10 +544,14 @@ def version_from_git(
 
+    filter_fxn = None
+    if tag_filter:
+        filter_fxn = load_tag_filter(tag_filter=tag_filter, package_name=package_name, root=root)
```
Please do not use abbreviations in variable names

dolfinus commented on tests/test_integration/test_tag_filter.py:
```diff
@@ -0,0 +1,142 @@
+    ],
+)
+def test_tag_filter(repo, create_config, tag_filter, version):
+    if tag_filter:
```
`tag_filter` is always set, `if` can be removed

dolfinus commented on tests/test_integration/test_tag_filter.py:
```diff
@@ -0,0 +1,142 @@
+        "product_y/1.1.10",
+        "product_z/foo/1.1.1",
+    ]
+
+    for i, tag in enumerate(tags_to_commit):
+        create_file(repo, commit=False)
+        dt = datetime.now() - timedelta(days=len(tags_to_commit) - i)
```
```suggestion

    for i, tag in enumerate(tags_to_commit):
        create_file(repo, commit=False)
        # create tags in the same order as in the list above
        dt = datetime.now() - timedelta(days=10) + timedelta(days=i)
```

vortechs2000 replied:
Will do

dolfinus commented on tests/test_integration/test_tag_filter.py:
```diff
@@ -0,0 +1,142 @@
+        dt = datetime.now() - timedelta(days=len(tags_to_commit) - i)
+        create_commit(repo, "Some commit", dt=dt)
+        commits[tag] = get_sha(repo)
+        time.sleep(1)
```
```suggestion
```

vortechs2000 replied:
This was copied in from the `test_tag.py` tests, and wasn't sure if it was needed.  I'll remove.

vortechs2000 replied:
Turns out it _was_ required, but I fixed the `create_commit()` function so it's no longer required in `test_tags.py` either.

dolfinus commented on tests/test_integration/test_tag_filter.py:
```diff
@@ -0,0 +1,142 @@
+    ]
+
+    for tag in tags_to_create:
+        create_tag(repo, tag, message="", commit=commits[tag])
```
Why not create tag immediately after the commit?

vortechs2000 replied:
I was just following the formatter test structure.

dolfinus commented on setuptools_git_versioning.py:
```diff
@@ -96,8 +98,16 @@ def get_branch_tags(*args, **kwargs) -> list[str]:
+) -> list[str]:
     tags = _exec(f"git tag --sort=-{sort_by} --merged", root=root)
+    if filter_fxn:
+        # pull the tags that don't start with tag_prefix out of the list
+        mytags = [n for n in list(map(filter_fxn, tags)) if n]
+        return mytags
```
```suggestion
        # get only tags matching the regexp
        return list(filter(filter_fxn, tags))
```

dolfinus commented on tests/test_integration/test_tag_filter.py:
```diff
@@ -0,0 +1,142 @@
+        ("1.0.0", "1.0.0", r"[\d.]+", None),
+        ("release/1.0.0", "0.0.1", r"[\d.]+", None),
+        ("unknown", "0.0.1", r"[\d.]+", None),
+        ("foo/bar/1.0.3-123", "1.0.3.123", r"foo/bar/.*", r"foo/bar/(?P<tag>.*)"),
```
Actually, you can pass `foo/bar/(?P<tag>.*)` to both `tag_filter` and `tag_formatter` - first will return True if tag is matched regexp (according to my previos comment), second will return version from parentheses

vortechs2000 replied:
Right, as long as you build the function/regexp for formatting, it can be used for filtering.  Vice versa isn't necessarily true because the filter only _requires_ a boolean response.  I'll add some verbiage for to recommend using the same regexp/callback as the formatter when using this.

vortechs2000 commented:
Ok @dolfinus, see what you think of that changeset.  Note that the `time.sleep(1)` calls in the test_tags.py tests are no longer required, so the tests should run a bit faster! 😁

dolfinus commented on tests/lib/util.py:
```diff
@@ -29,7 +29,7 @@ def rand_sha() -> str:
 
 def execute(cwd: str | os.PathLike, cmd: str, **kwargs) -> str:
-    log.info(f"Executing '{cmd}' at '{cwd}'")
+    log.critical(f"Executing '{cmd}' at '{cwd}'")
```
```suggestion
    log.info(f"Executing '{cmd}' at '{cwd}'")
```

dolfinus commented on tests/lib/util.py:
```diff
@@ -48,9 +48,27 @@ def create_commit(
         options += f"--date {dt.isoformat()}"
-    return execute(cwd, f'git commit -m "{message}" {options}', **kwargs)
+        # The committer date is what is used to determine sort order for tags, etc
+        os.environ["GIT_COMMITTER_DATE"] = f"--date {dt.isoformat()}"
```
`GIT_COMMITER_DATE` should contain date, not `--date $value`
https://git-scm.com/docs/git-commit/2.24.0#_date_formats

vortechs2000 replied:
Yes, yes it should.  Now I'm doubting the testing I did.  So confused...

dolfinus commented on tests/lib/util.py:
```diff
@@ -48,9 +48,27 @@ def create_commit(
+
+    # Return committer date env var to prior value if set
+    if dt is not None:
+        if original_committer_date is None:
```
If `execute()` will raise an exception, environment variable value will not be reverted. Use try.finally

dolfinus commented on tests/test_integration/test_tag.py:
```diff
@@ -299,16 +297,8 @@ def test_tag_sort_by_tag_date(repo, create_config, message):
-        assert "1.1.10" in get_version(repo)
-    else:
-        assert get_version(repo).startswith("1.1")
-        # the result is totally random because annotaged tags have no such field at all
-        # https://github.com/dolfinus/setuptools-git-versioning/issues/23
+    assert get_version(repo).startswith("1.1")
```
Please revert this. `GIT_COMMITER_DATE` has nothing in common with tag creation date

vortechs2000 replied:
K - IIRC, the tests weren't passing with the GIT_COMMITTER_DATE set, but since it was set incorrectly, I'll see what happens when it's fixed.

vortechs2000 replied:
Yep, works fine with the ``GIT_COMMITTER_DATE`` fix and the revert.

dolfinus commented on tests/test_integration/test_tag_filter.py:
```diff
@@ -0,0 +1,130 @@
+        create_file(repo, commit=False)
+        dt = datetime.now() - timedelta(days=10) + timedelta(days=i)
+        create_commit(repo, "Some commit", dt=dt)
+        commits[tag] = get_sha(repo)
+        create_tag(repo, tag, message="", commit=commits[tag])
```
```suggestion
        create_tag(repo, tag, commit=get_sha(repo))
```

vortechs2000 commented:
@dolfinus - ok, addressed your comments on the latest patchset, and I think the code conforms to the changes you just made in master.  

The code coverage fails because the tests don't hit the case in ``_load_filter_tag`` where the tag_filter passed in is already a direct callable.  That code mirrors the code in the other ``_load_*`` functions that handle callables, so I _think_ we want it there.  Thoughts?

dolfinus commented:
Thank you, I'll check what can be done with coverage
